File: src/cli/proxy.ts
import pc from "picocolors";
import * as p from "@clack/prompts";
import { spawn } from "node:child_process";
import type { ClaudeCodeProfile } from "../types.js";
import { getProfile } from "../config/manager.js";
import { spawnClaudeWithEnv, applyClaudeProfile } from "./claude.js";
import {
  DEFAULT_PROXY_HOST,
  DEFAULT_PROXY_PORT,
  SWIXTER_PROXY_AUTH_TOKEN,
} from "../constants/proxy.js";
import { buildClaudeProxyMarkerModels } from "../utils/model-helper.js";
import { EXIT_CODES } from "../constants/formatting.js";
import {
  startProxyServer,
  stopProxyServer,
  getProxyStatus,
  listProxyInstances,
} from "../proxy/server.js";
import { getActiveGroup, getGroup } from "../groups/manager.js";

export function resolveProxyRuntimeBinding(input: {
  groupName: string;
  requestedPort?: number;
  allInstances: ReturnType<typeof listProxyInstances>;
}): { host: string; port: number; reuseExisting: boolean; reuseInstanceId?: string } {
  const occupiedPorts = new Set(
    input.allInstances.filter((s) => s.running).map((s) => s.port)
  );

  // Explicit port requested
  if (input.requestedPort) {
    return {
      host: DEFAULT_PROXY_HOST,
      port: input.requestedPort,
      reuseExisting: false,
    };
  }

  // Check if any existing instance already serves this group
  const existing = input.allInstances.find(
    (s) => s.running && s.groupName === input.groupName
  );
  if (existing) {
    return {
      host: existing.host,
      port: existing.port,
      reuseExisting: true,
      reuseInstanceId: existing.instanceId,
    };
  }

  // Find next available port starting from default
  let port = DEFAULT_PROXY_PORT;
  while (occupiedPorts.has(port)) {
    port++;
  }

  return {
    host: DEFAULT_PROXY_HOST,
    port,
    reuseExisting: false,
  };
}

export function buildClaudeProxyEnv(
  env: NodeJS.ProcessEnv,
  port: number
): NodeJS.ProcessEnv {
  const nextEnv = { ...env };
  nextEnv.ANTHROPIC_API_BASE = `http://${DEFAULT_PROXY_HOST}:${port}`;
  nextEnv.ANTHROPIC_AUTH_TOKEN = SWIXTER_PROXY_AUTH_TOKEN;
  delete nextEnv.ANTHROPIC_API_KEY;
  return nextEnv;
}

export function buildCoderProxyEnv(
  coder: string,
  baseEnv: NodeJS.ProcessEnv,
  port: number
): NodeJS.ProcessEnv {
  if (coder === "claude") {
    return buildClaudeProxyEnv(baseEnv, port);
  }

  const env = { ...baseEnv };

  if (coder === "qwen") {
    env.ANTHROPIC_API_BASE = `http://${DEFAULT_PROXY_HOST}:${port}`;
    env.ANTHROPIC_API_KEY = "dummy";
    delete env.ANTHROPIC_AUTH_TOKEN;
  } else if (coder === "codex") {
    env.OPENAI_API_BASE = `http://${DEFAULT_PROXY_HOST}:${port}`;
    env.OPENAI_API_KEY = "dummy";
  }

  return env;
}

export async function waitForProxyHealth(host: string, port: number, attempts = 10): Promise<boolean> {
  const url = `http://${host}:${port}/health`;

  for (let i = 0; i < attempts; i++) {
    try {
      const response = await fetch(url);
      if (response.ok) {
        return true;
      }
    } catch {
      // Continue polling until attempts are exhausted.
    }

    await new Promise(resolve => setTimeout(resolve, 100));
  }

  return false;
}

export async function waitForProxyRuntime(host: string, port: number, attempts = 10): Promise<boolean> {
  for (let i = 0; i < attempts; i++) {
    const instances = listProxyInstances();
    const match = instances.find((s) => s.running && s.host === host && s.port === port);
    if (match) return true;

    await new Promise(resolve => setTimeout(resolve, 100));
  }

  return false;
}

export async function handleProxyCommand(args: string[]): Promise<void> {
  const subcommand = args[0];

  switch (subcommand) {
    case "start":
      await cmdStart(args.slice(1));
      break;
    case "stop":
      await cmdStop(args.slice(1));
      break;
    case "run":
      await cmdRun(args.slice(1));
      break;
    case "status":
      await cmdStatus();
      break;
    default:
      if (!subcommand || subcommand === "status") {
        await cmdStatus();
      } else {
        console.log(pc.red(`Unknown subcommand: ${subcommand}`));
        console.log(proxyHelp());
        process.exit(1);
      }
  }
}

async function cmdStart(args: string[]): Promise<void> {
  let groupName: string | undefined;
  let port: number | undefined;
  let host = "127.0.0.1";
  let timeout = 3000000;
  let daemon = false;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--group" && args[i + 1]) {
      groupName = args[i + 1];
      i++;
    } else if (args[i] === "--port" && args[i + 1]) {
      port = parseInt(args[i + 1], 10);
      i++;
    } else if (args[i] === "--host" && args[i + 1]) {
      host = args[i + 1];
      i++;
    } else if (args[i] === "--timeout" && args[i + 1]) {
      timeout = parseInt(args[i + 1], 10);
      i++;
    } else if (args[i] === "--daemon") {
      daemon = true;
    }
  }

  if (groupName) {
    const group = await getGroup(groupName);
    if (!group) {
      console.log(pc.red(`Group "${groupName}" not found`));
      process.exit(EXIT_CODES.notFound);
      return;
    }
  }

  // Check if default instance is already running
  const defaultStatus = getProxyStatus("default");
  if (defaultStatus.running) {
    console.log(pc.yellow(`Default proxy already running on ${defaultStatus.host}:${defaultStatus.port}`));
    return;
  }

  if (!groupName) {
    const activeGroup = await getActiveGroup();
    if (activeGroup) {
      groupName = activeGroup.name;
      console.log(pc.dim(`Using default group: ${groupName}`));
    }
  }

  const resolvedPort = port || DEFAULT_PROXY_PORT;

  if (daemon) {
    await cmdStartDaemon({ host, port: resolvedPort, timeout, groupName });
  } else {
    await cmdStartBlocking({ host, port: resolvedPort, timeout, groupName });
  }
}

async function cmdStartBlocking(config: { host: string; port: number; timeout: number; groupName?: string }): Promise<void> {
  console.log(pc.cyan("Starting proxy server..."));

  await startProxyServer({
    instanceId: "default",
    type: "service",
    host: config.host,
    port: config.port,
    timeout: config.timeout,
    groupName: config.groupName,
  });

  console.log();
  console.log(pc.green("✓ Proxy server started"));
  console.log(`  Instance: default (service)`);
  console.log(`  Address: ${config.host}:${config.port}`);
  console.log(`  Timeout: ${config.timeout}ms`);
  console.log(pc.dim(`  Group: ${config.groupName || "none"}`));
  console.log(pc.dim(`  Endpoints:`));
  console.log(pc.dim(`    - /v1/chat/completions (OpenAI)`));
  console.log(pc.dim(`    - /v1/messages (Anthropic)`));
  console.log(pc.dim(`    - /v1/responses (Anthropic)`));
  console.log(pc.dim(`    - /anthropic/* (Anthropic)`));
  console.log(pc.dim(`    - /health (Health check)`));
  console.log();
  console.log(pc.dim("Press Ctrl+C to stop"));
}

async function cmdStartDaemon(config: { host: string; port: number; timeout: number; groupName?: string }): Promise<void> {
  const defaultStatus = getProxyStatus("default");
  if (defaultStatus.running) {
    console.log(pc.yellow(`Default proxy already running on ${defaultStatus.host}:${defaultStatus.port}`));
    return;
  }

  console.log(pc.cyan("Starting proxy server in background..."));

  const args = [
    "proxy",
    "start",
    "--host", config.host,
    "--port", config.port.toString(),
    "--timeout", config.timeout.toString(),
  ];

  if (config.groupName) {
    args.push("--group", config.groupName);
  }

  const child = spawn(process.execPath, [process.argv[1], ...args], {
    detached: true,
    stdio: "ignore",
  });

  child.unref();

  const isHealthy = await waitForProxyHealth(config.host, config.port);
  const hasRuntime = isHealthy
    ? await waitForProxyRuntime(config.host, config.port)
    : false;

  if (isHealthy && hasRuntime) {
    console.log(pc.green("✓ Proxy server started in background"));
    console.log(`  Instance: default (service)`);
    console.log(`  Address: ${config.host}:${config.port}`);
    console.log(pc.dim(`  Use "swixter proxy status" to check status`));
    console.log(pc.dim(`  Use "swixter proxy stop" to stop`));
  } else {
    console.log(pc.red("✗ Failed to start proxy server in background"));
    process.exit(1);
  }
}

async function cmdStop(args: string[]): Promise<void> {
  // Support: swixter proxy stop [instanceId]
  const instanceId = args[0] || "default";
  const status = getProxyStatus(instanceId);

  if (!status.running) {
    console.log(pc.yellow(`Proxy instance "${instanceId}" is not running`));
    return;
  }

  await stopProxyServer(instanceId);
  console.log(pc.green(`✓ Proxy instance "${instanceId}" stopped`));
}

async function cmdRun(args: string[]): Promise<void> {
  const doubleDash = args.indexOf("--");
  let groupArgs: string[] = [];
  let coderArgs: string[] = [];

  if (doubleDash >= 0) {
    groupArgs = args.slice(0, doubleDash);
    coderArgs = args.slice(doubleDash + 1);
  } else {
    groupArgs = args;
  }

  let groupName: string | undefined;
  let requestedPort: number | undefined;
  for (let i = 0; i < groupArgs.length; i++) {
    if (groupArgs[i] === "--group" && groupArgs[i + 1]) {
      groupName = groupArgs[i + 1];
      i++;
    } else if (groupArgs[i] === "--port" && groupArgs[i + 1]) {
      requestedPort = parseInt(groupArgs[i + 1], 10);
      i++;
    }
  }

  if (!groupName) {
    const activeGroup = await getActiveGroup();
    groupName = activeGroup?.name;
  }

  if (!groupName) {
    console.log(pc.red("No group specified and no default group set"));
    console.log(pc.dim("Use --group or create a default group first"));
    return;
  }

  const group = await getGroup(groupName);
  if (!group) {
    console.log(pc.red(`Group "${groupName}" not found`));
    process.exit(EXIT_CODES.notFound);
    return;
  }

  console.log(pc.cyan("Starting proxy server..."));
  const allInstances = listProxyInstances();
  const runtimeBinding = resolveProxyRuntimeBinding({
    groupName,
    requestedPort,
    allInstances,
  });

  const instanceId = `run-${runtimeBinding.port}`;

  if (!runtimeBinding.reuseExisting) {
    await startProxyServer({
      instanceId,
      type: "run",
      host: runtimeBinding.host,
      port: runtimeBinding.port,
      groupName,
    });
  }

  const coder = coderArgs[0];
  if (!coder) {
    console.log(pc.red("Coder command required after --"));
    console.log(pc.dim("Example: swixter proxy run -- claude"));
    await stopProxyServer(instanceId);
    return;
  }

  const env = buildCoderProxyEnv(coder, process.env, runtimeBinding.port);

  console.log(pc.green(`✓ Running: ${coder} ${coderArgs.slice(1).join(" ")}`));
  console.log(pc.dim(`  Instance: ${instanceId} (run)`));
  console.log(pc.dim(`  Proxy: ${runtimeBinding.host}:${runtimeBinding.port}`));

  if (coder === "claude") {
    const firstProfile = group.profiles[0] ? await getProfile(group.profiles[0]) : null;
    const proxyProfile: ClaudeCodeProfile = {
      name: `proxy-${groupName}`,
      providerId: "anthropic",
      apiKey: "",
      authToken: SWIXTER_PROXY_AUTH_TOKEN,
      baseURL: `http://${runtimeBinding.host}:${runtimeBinding.port}`,
      models: firstProfile ? buildClaudeProxyMarkerModels(firstProfile) : undefined,
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    };

    await applyClaudeProfile(proxyProfile);
    await spawnClaudeWithEnv(coderArgs.slice(1), env as Record<string, string>, {
      onExit: async () => {
        await stopProxyServer(instanceId);
      },
    });
    return;
  }

  const child = spawn(coder, coderArgs.slice(1), {
    env,
    stdio: "inherit",
  });

  child.on("exit", async (code) => {
    await stopProxyServer(instanceId);
    process.exit(code || 0);
  });

  process.on("SIGINT", async () => {
    child.kill("SIGINT");
    await stopProxyServer(instanceId);
    process.exit(1);
  });
}

async function cmdStatus(): Promise<void> {
  const instances = listProxyInstances();

  console.log();
  console.log(pc.bold("Proxy Status:"));
  console.log();

  const running = instances.filter((s) => s.running);

  if (running.length === 0) {
    console.log(`  ${pc.red("●")} No proxy instances running`);
    console.log();
    console.log(pc.dim("  Start with: swixter proxy start"));
    return;
  }

  for (const status of running) {
    const typeLabel = status.type === "service" ? pc.cyan("service") : pc.yellow("run");
    console.log(`  ${pc.green("●")} ${pc.bold(status.instanceId)} (${typeLabel})`);
    console.log(`    Address: ${status.host}:${status.port}`);
    console.log(`    Group: ${status.groupName || "none"}`);
    console.log(`    Requests: ${status.requestCount} | Errors: ${status.errorCount}`);
    if (status.startTime) {
      console.log(`    Started: ${status.startTime}`);
    }
    console.log();
  }
}

function proxyHelp(): string {
  return `
${pc.bold("Swixter Proxy Commands")}

${pc.bold("Usage:")}
  ${pc.green("swixter proxy <command> [options]")}

${pc.bold("Commands:")}
  ${pc.cyan("start")}              Start proxy server (default instance)
  ${pc.cyan("stop")} [instanceId]  Stop proxy instance (default: "default")
  ${pc.cyan("run")}                Start proxy and run coder with env vars
  ${pc.cyan("status")}             Show all proxy instances

${pc.bold("Options:")}
  ${pc.dim("--group <name>")}      Use specified group
  ${pc.dim("--port <port>")}       Proxy port (default: 15721)
  ${pc.dim("--host <host>")}       Proxy host (default: 127.0.0.1)
  ${pc.dim("--timeout <ms>")}      Request timeout in ms (default: 3000000)
  ${pc.dim("--daemon")}             Run in background (detached)

${pc.bold("Examples:")}
  ${pc.green("swixter proxy start")}
  ${pc.green("swixter proxy start --daemon")}
  ${pc.green("swixter proxy start --group my-group")}
  ${pc.green("swixter proxy run --group my-group -- claude")}
  ${pc.green("swixter proxy stop")}
  ${pc.green("swixter proxy stop run-15722")}
  ${pc.dim("  # Stop a specific run instance")}
`;
}

File: packages/cli/crates/swixter/src/commands/proxy.rs
use std::process::Stdio;
use std::time::Duration;

use swixter_core::adapters::get_adapter;
use swixter_core::coder::AdapterKind;
use swixter_core::config::ConfigManager;
use swixter_core::types::Profile;
use swixter_proxy::registry;
use swixter_proxy::server;
use swixter_proxy::types::{InstanceKind, ProxyServerConfig, ProxyStatus};
use swixter_proxy::{DEFAULT_PROXY_HOST, DEFAULT_PROXY_PORT, SWIXTER_PROXY_AUTH_TOKEN};

use crate::cli::{ProxyArgs, ProxyCommand, ProxyRunArgs, ProxyStartArgs};
use crate::{EXIT_GENERAL, EXIT_INVALID_ARG, EXIT_NOT_FOUND, EXIT_SUCCESS};

pub fn dispatch(args: ProxyArgs) -> i32 {
    match args.command {
        ProxyCommand::Start(a) => cmd_start(a),
        ProxyCommand::Stop { instance_id } => cmd_stop(instance_id.as_deref().unwrap_or("default")),
        ProxyCommand::Status => cmd_status(),
        ProxyCommand::Run(a) => cmd_run(a),
    }
}

fn runtime() -> tokio::runtime::Runtime {
    tokio::runtime::Builder::new_multi_thread()
        .enable_all()
        .build()
        .expect("tokio runtime")
}

/// group/profile 互斥 + 存在性校验（start/run 共用，TS cli/proxy.ts 同款）
fn validate_group_profile(
    mgr: &ConfigManager,
    group: Option<&str>,
    profile: Option<&str>,
) -> Result<(), i32> {
    if group.is_some() && profile.is_some() {
        eprintln!("Cannot specify both --group and --profile");
        return Err(EXIT_INVALID_ARG);
    }
    if let Some(g) = group {
        if mgr
            .config()
            .groups
            .values()
            .all(|x| x.name != *g && x.id != *g)
        {
            eprintln!("Group \"{g}\" not found");
            return Err(EXIT_NOT_FOUND);
        }
    }
    if let Some(p) = profile {
        if mgr.get_profile(p).is_none() {
            eprintln!("Profile \"{p}\" not found");
            return Err(EXIT_NOT_FOUND);
        }
    }
    Ok(())
}

fn cmd_start(a: ProxyStartArgs) -> i32 {
    let mgr = ConfigManager::load();
    if let Err(code) = validate_group_profile(&mgr, a.group.as_deref(), a.profile.as_deref()) {
        return code;
    }
    let default_status = registry::get_proxy_status("default");
    if default_status.running {
        println!(
            "Default proxy already running on {}:{}",
            default_status.host, default_status.port
        );
        return EXIT_SUCCESS;
    }
    // 未指定 group/profile 时用 active group
    let mut group = a.group.clone();
    if group.is_none() && a.profile.is_none() {
        if let Some(g) = mgr
            .config()
            .active_group
            .as_ref()
            .and_then(|id| mgr.config().groups.get(id))
        {
            println!("Using default group: {}", g.name);
            group = Some(g.name.clone());
        }
    }
    if a.daemon {
        cmd_start_daemon(&a, group.as_deref())
    } else {
        let config = ProxyServerConfig {
            instance_id: "default".into(),
            kind: InstanceKind::Service,
            host: a.host.clone(),
            port: a.port,
            timeout: Duration::from_millis(a.timeout),
            group_name: group,
            profile_name: a.profile.clone(),
            config_path: None,
        };
        runtime().block_on(async move {
            match server::start_proxy_server(config).await {
                Ok(s) => {
                    println!("✓ Proxy server started");
                    println!("  Instance: default (service)");
                    println!("  Address: {}:{}", s.host, s.port);
                    println!("  Press Ctrl+C to stop");
                    let _ = tokio::signal::ctrl_c().await;
                    server::stop_in_process_instance("default").await;
                    EXIT_SUCCESS
                }
                Err(e) => {
                    eprintln!("✗ {e}");
                    EXIT_GENERAL
                }
            }
        })
    }
}

fn cmd_start_daemon(a: &ProxyStartArgs, group: Option<&str>) -> i32 {
    // spawn 自身 detached：proxy start 同参（去掉 --daemon），stdio 全 null
    let exe = std::env::current_exe().expect("current exe");
    let mut cmd = std::process::Command::new(exe);
    cmd.args([
        "proxy",
        "start",
        "--host",
        &a.host,
        "--port",
        &a.port.to_string(),
        "--timeout",
        &a.timeout.to_string(),
    ]);
    if let Some(g) = group {
        cmd.args(["--group", g]);
    }
    if let Some(p) = &a.profile {
        cmd.args(["--profile", p]);
    }
    cmd.stdin(Stdio::null())
        .stdout(Stdio::null())
        .stderr(Stdio::null());
    #[cfg(unix)]
    unsafe {
        use std::os::unix::process::CommandExt;
        cmd.pre_exec(|| {
            libc::setsid();
            Ok(())
        });
    }
    #[cfg(windows)]
    {
        use std::os::windows::process::CommandExt;
        const CREATE_NO_WINDOW: u32 = 0x08000000;
        const DETACHED_PROCESS: u32 = 0x00000008;
        cmd.creation_flags(CREATE_NO_WINDOW | DETACHED_PROCESS);
    }
    if let Err(e) = cmd.spawn() {
        eprintln!("✗ Failed to spawn daemon: {e}");
        return EXIT_GENERAL;
    }
    // 轮询 /health（10×100ms）+ registry runtime（10×100ms）
    let ok = runtime().block_on(async {
        for _ in 0..10 {
            if server::health_check(&a.host, a.port).await {
                for _ in 0..10 {
                    let found = registry::list_proxy_instances()
                        .iter()
                        .any(|s| s.running && s.host == a.host && s.port == a.port);
                    if found {
                        return true;
                    }
                    tokio::time::sleep(Duration::from_millis(100)).await;
                }
                return true; // health 通过即视为启动（runtime 轮询失败不致命）
            }
            tokio::time::sleep(Duration::from_millis(100)).await;
        }
        false
    });
    if ok {
        println!("✓ Proxy server started in background");
        println!("  Address: {}:{}", a.host, a.port);
        EXIT_SUCCESS
    } else {
        eprintln!("✗ Failed to start proxy server in background");
        EXIT_GENERAL
    }
}

fn cmd_stop(instance_id: &str) -> i32 {
    let status = registry::get_proxy_status(instance_id);
    if !status.running {
        println!("Proxy instance \"{instance_id}\" is not running");
        return EXIT_SUCCESS;
    }
    runtime().block_on(async {
        if server::stop_in_process_instance(instance_id).await {
            println!("✓ Proxy instance \"{instance_id}\" stopped");
            return EXIT_SUCCESS;
        }
        // 跨进程：按 registry pid 发信号 kill（决策点 3，改进 TS 只删条目的现状）
        if let Some(pid) = status.pid {
            registry::terminate_process(pid);
        }
        registry::remove_instance(instance_id);
        println!("✓ Proxy instance \"{instance_id}\" stopped");
        EXIT_SUCCESS
    })
}

fn cmd_status() -> i32 {
    let instances = registry::list_proxy_instances();
    let running: Vec<_> = instances.iter().filter(|s| s.running).collect();
    println!();
    println!("Proxy Status:");
    println!();
    if running.is_empty() {
        println!("  ● No proxy instances running");
        println!();
        println!("  Start with: swixter proxy start");
        return EXIT_SUCCESS;
    }
    for s in running {
        let kind = if s.kind == InstanceKind::Service {
            "service"
        } else {
            "run"
        };
        println!("  ● {} ({kind})", s.instance_id);
        println!("    Address: {}:{}", s.host, s.port);
        println!("    Group: {}", s.group_name.as_deref().unwrap_or("none"));
        println!(
            "    Profile: {}",
            s.profile_name.as_deref().unwrap_or("none")
        );
        println!(
            "    Requests: {} | Errors: {}",
            s.request_count, s.error_count
        );
        if let Some(t) = &s.start_time {
            println!("    Started: {t}");
        }
        println!();
    }
    EXIT_SUCCESS
}

pub struct RuntimeBinding {
    pub host: String,
    pub port: u16,
    pub reuse_existing: bool,
    pub reuse_instance_id: Option<String>,
}

/// TS: resolveProxyRuntimeBinding
pub fn resolve_proxy_runtime_binding(
    group_name: Option<&str>,
    profile_name: Option<&str>,
    requested_port: Option<u16>,
    all_instances: &[ProxyStatus],
) -> RuntimeBinding {
    if let Some(port) = requested_port {
        return RuntimeBinding {
            host: DEFAULT_PROXY_HOST.into(),
            port,
            reuse_existing: false,
            reuse_instance_id: None,
        };
    }
    if let Some(existing) = all_instances.iter().find(|s| {
        s.running
            && ((group_name.is_some() && s.group_name.as_deref() == group_name)
                || (profile_name.is_some() && s.profile_name.as_deref() == profile_name))
    }) {
        return RuntimeBinding {
            host: existing.host.clone(),
            port: existing.port,
            reuse_existing: true,
            reuse_instance_id: Some(existing.instance_id.clone()),
        };
    }
    let occupied: std::collections::HashSet<u16> = all_instances
        .iter()
        .filter(|s| s.running)
        .map(|s| s.port)
        .collect();
    let mut port = DEFAULT_PROXY_PORT;
    while occupied.contains(&port) {
        port += 1;
    }
    RuntimeBinding {
        host: DEFAULT_PROXY_HOST.into(),
        port,
        reuse_existing: false,
        reuse_instance_id: None,
    }
}

/// TS: buildCoderProxyEnv
pub fn build_coder_proxy_env(
    coder: &str,
    base: &[(String, String)],
    port: u16,
) -> Vec<(String, String)> {
    let base_url = format!("http://{DEFAULT_PROXY_HOST}:{port}");
    let mut env: Vec<(String, String)> = base.to_vec();
    let set = |env: &mut Vec<(String, String)>, k: &str, v: &str| {
        env.retain(|(key, _)| key != k);
        env.push((k.to_string(), v.to_string()));
    };
    let unset = |env: &mut Vec<(String, String)>, k: &str| env.retain(|(key, _)| key != k);
    match coder {
        "claude" => {
            set(&mut env, "ANTHROPIC_API_BASE", &base_url);
            set(&mut env, "ANTHROPIC_AUTH_TOKEN", SWIXTER_PROXY_AUTH_TOKEN);
            unset(&mut env, "ANTHROPIC_API_KEY");
        }
        "qwen" => {
            set(&mut env, "ANTHROPIC_API_BASE", &base_url);
            set(&mut env, "ANTHROPIC_API_KEY", "dummy");
            unset(&mut env, "ANTHROPIC_AUTH_TOKEN");
        }
        "codex" => {
            set(&mut env, "OPENAI_API_BASE", &base_url);
            set(&mut env, "OPENAI_API_KEY", "dummy");
        }
        _ => {}
    }
    env
}

fn cmd_run(a: ProxyRunArgs) -> i32 {
    let mgr = ConfigManager::load();
    if let Err(code) = validate_group_profile(&mgr, a.group.as_deref(), a.profile.as_deref()) {
        return code;
    }
    // 都未指定 → active group；仍无 → 报错提示
    let mut group = a.group.clone();
    if group.is_none() && a.profile.is_none() {
        group = mgr
            .config()
            .active_group
            .as_ref()
            .and_then(|id| mgr.config().groups.get(id))
            .map(|g| g.name.clone());
    }
    if group.is_none() && a.profile.is_none() {
        eprintln!("No group or profile specified, and no default group set");
        eprintln!("Use --group, --profile, or create a default group first");
        return EXIT_GENERAL;
    }

    let instances = registry::list_proxy_instances();
    let binding =
        resolve_proxy_runtime_binding(group.as_deref(), a.profile.as_deref(), a.port, &instances);
    let instance_id = format!("run-{}", binding.port);
    // 有意偏差（Global Constraints）：复用实例时 coder 退出不停该实例；TS 会误停
    let started_by_us = !binding.reuse_existing;
    if let Some(id) = &binding.reuse_instance_id {
        println!("✓ Reusing running instance: {id}");
    }

    let coder_args = a.args.clone();
    let Some(coder) = coder_args.first().cloned() else {
        eprintln!("Coder command required after --");
        eprintln!("Example: swixter proxy run -- claude");
        return EXIT_GENERAL;
    };
    // claude：proxy profile + marker models 写入 ~/.claude/settings.json（TS applyClaudeProfile 路径）
    if coder == "claude" {
        let target = a
            .profile
            .as_ref()
            .and_then(|n| mgr.get_profile(n))
            .or_else(|| {
                group
                    .as_ref()
                    .and_then(|g| {
                        mgr.config()
                            .groups
                            .values()
                            .find(|x| x.name == *g || x.id == *g)
                    })
                    .and_then(|g| g.profiles.first())
                    .and_then(|n| mgr.get_profile(n))
            });
        let proxy_profile = Profile {
            name: format!(
                "proxy-{}",
                a.profile
                    .as_deref()
                    .or(group.as_deref())
                    .unwrap_or("default")
            ),
            provider_id: "anthropic".into(),
            api_key: String::new(),
            auth_token: Some(SWIXTER_PROXY_AUTH_TOKEN.into()),
            base_url: Some(format!("http://{}:{}", binding.host, binding.port)),
            models: target.and_then(swixter_core::model::build_claude_proxy_marker_models),
            created_at: swixter_core::types::now_iso(),
            updated_at: swixter_core::types::now_iso(),
            ..Default::default()
        };
        let adapter = get_adapter(AdapterKind::Claude);
        if let Err(e) = adapter.apply(&proxy_profile, None) {
            eprintln!("✗ Failed to apply claude proxy profile: {e}");
            return EXIT_GENERAL;
        }
    }

    let base_env: Vec<(String, String)> = std::env::vars().collect();
    let env = build_coder_proxy_env(&coder, &base_env, binding.port);

    runtime().block_on(async move {
        if started_by_us {
            let config = ProxyServerConfig {
                instance_id: instance_id.clone(),
                kind: InstanceKind::Run,
                host: binding.host.clone(),
                port: binding.port,
                // TS run 不传 timeout → forwarder 默认
                timeout: Duration::from_millis(swixter_proxy::DEFAULT_TIMEOUT_MS),
                group_name: group,
                profile_name: a.profile,
                config_path: None,
            };
            if let Err(e) = server::start_proxy_server(config).await {
                eprintln!("✗ {e}");
                return EXIT_GENERAL;
            }
        }
        println!("✓ Running: {} {}", coder, coder_args[1..].join(" "));
        println!("  Proxy: {}:{}", binding.host, binding.port);

        // env 已是完整环境（含删除项处理），必须 env_clear 才能真正删掉 ANTHROPIC_API_KEY 等
        let mut child = match tokio::process::Command::new(&coder)
            .args(&coder_args[1..])
            .env_clear()
            .envs(env)
            .stdin(Stdio::inherit())
            .stdout(Stdio::inherit())
            .stderr(Stdio::inherit())
            .spawn()
        {
            Ok(c) => c,
            Err(e) => {
                eprintln!("✗ Failed to launch {coder}: {e}");
                if started_by_us {
                    server::stop_in_process_instance(&instance_id).await;
                }
                return EXIT_GENERAL;
            }
        };

        // coder 退出 → 停实例、透传退出码；Ctrl+C → 转发 + 停 + exit 1
        let code = tokio::select! {
            status = child.wait() => status.map(|s| s.code().unwrap_or(0)).unwrap_or(0),
            _ = tokio::signal::ctrl_c() => {
                let _ = child.kill().await;
                if started_by_us {
                    server::stop_in_process_instance(&instance_id).await;
                }
                return 1;
            }
        };
        if started_by_us {
            server::stop_in_process_instance(&instance_id).await;
        }
        code
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn coder_env_injection() {
        let base = vec![
            ("ANTHROPIC_API_KEY".to_string(), "old".to_string()),
            ("PATH".to_string(), "/bin".to_string()),
        ];
        let env = build_coder_proxy_env("claude", &base, 15721);
        assert!(env
            .iter()
            .any(|(k, v)| k == "ANTHROPIC_API_BASE" && v == "http://127.0.0.1:15721"));
        assert!(env
            .iter()
            .any(|(k, v)| k == "ANTHROPIC_AUTH_TOKEN" && v == "swixter-local-proxy"));
        assert!(!env.iter().any(|(k, _)| k == "ANTHROPIC_API_KEY")); // 删除
        let env = build_coder_proxy_env("qwen", &base, 15721);
        assert!(env
            .iter()
            .any(|(k, v)| k == "ANTHROPIC_API_KEY" && v == "dummy"));
        let env = build_coder_proxy_env("codex", &base, 15721);
        assert!(env
            .iter()
            .any(|(k, v)| k == "OPENAI_API_BASE" && v == "http://127.0.0.1:15721"));
        assert!(env
            .iter()
            .any(|(k, v)| k == "OPENAI_API_KEY" && v == "dummy"));
    }

    #[test]
    fn runtime_binding_reuse_and_port_scan() {
        let running = |id: &str, port: u16, group: Option<&str>| ProxyStatus {
            instance_id: id.into(),
            running: true,
            port,
            group_name: group.map(Into::into),
            ..Default::default()
        };
        let instances = vec![
            running("default", 15721, Some("g1")),
            running("run-15722", 15722, None),
        ];
        // 显式 port 直接生效
        let b = resolve_proxy_runtime_binding(None, None, Some(16000), &instances);
        assert_eq!(b.port, 16000);
        assert!(!b.reuse_existing);
        // 同 group 复用
        let b = resolve_proxy_runtime_binding(Some("g1"), None, None, &instances);
        assert!(b.reuse_existing);
        assert_eq!(b.port, 15721);
        // 否则从 15721 起找空位
        let b = resolve_proxy_runtime_binding(Some("g2"), None, None, &instances);
        assert_eq!(b.port, 15723);
        assert!(!b.reuse_existing);
    }
}
